core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+
+        // Initialize ckeditor.
+        var ck = CKEDITOR.inline(el[0].children[0].children[1], {
+          extraPlugins: 'widget,lineutils,sharedspace,' + pluginNames,
+          startupFocus: true,
+          extraAllowedContent: componentFilterRules + wrapperFilterRules,
+          sharedSpaces: {
+            top: el[0].children[0].children[0]
+          },
+          skin: 'bootstrapck,/third_party/static/ckeditor-bootstrapck/',
+          toolbar: [
+            {
+              name: 'history',
+              items: ['Undo', '-', 'Redo']
+            },
+            {
+              name: 'basicstyles',
+              items: ['Bold', '-', 'Italic', '-', 'RemoveFormat']
+            },
+            {
+              name: 'paragraph',
+              items: [
+                'NumberedList', '-',
+                'BulletedList', '-',
+                'Outdent', '-',
+                'Indent'
+              ]
+            },
+            {
+              name: 'rtecomponents',
+              items: buttonNames
+            }
+          ]
+        });
+
+        // A RegExp for matching rich text components.
+        var componentRe = (
+          /(<(oppia-noninteractive-(.+?))\b[^>]*>)[\s\S]*?<\/\2>/g
+        );
+
+        // Before data is loaded into CKEditor, we need to wrap every rte
+        // component in a span (inline) or div (block).
+        // For block elements, we add an overlay div as well.
+        var wrapComponents = function(html) {
+          if (html === undefined) {
+            return html;
+          }
+          return html.replace(componentRe, function(match, p1, p2, p3) {
+            if (rteHelperService.isInlineComponent(p3)) {
+              return '<span type="oppia-noninteractive-' + p3 + '">' +
+                    match + '</span>';
+            } else {
+              return '<div type="oppia-noninteractive-' + p3 + '"' +
+                     'class="component-container">' + match +
+                     '<div class="component-overlay"></div></div>';
+            }
+          });
+        };
+
+        ck.on('instanceReady', function() {
+          // Set the icons for each toolbar button.
+          names.forEach(function(name, index) {
+            var icon = icons[index];
+            var upperCasedName = name.charAt(0).toUpperCase() + name.slice(1);
+            $('.cke_button__oppia' + name)
+               .css('background-image', 'url("' + icon + '")')
+               .css('background-position', 'center')
+               .css('background-repeat', 'no-repeat');
+          });
+          ck.setData(wrapComponents(ngModel.$viewValue));
+        });
+
+        // Angular rendering of components confuses CKEditor's undo system, so
+        // we hide all of that stuff away from CKEditor.
+        ck.on('getSnapshot', function(event) {
+          if (event.data === undefined) {
+            return;
+          }
+          event.data = event.data.replace(componentRe, function(match, p1, p2) {
```

Comment:
Just so I understand, where's the corresponding logic that reinstates the full tags when a snapshot is subsequently reinstated?